### PR TITLE
Support windows

### LINKED ISSUE #134: Installation on windows
Hi,

I really want to give stack_master a go however I am having trouble installing the gem on windows. I am running gem 2.6.7 and it pulls down all the dependencies however when it gets to "building native extensions"  it errors. I have looked at the logs but I am unable to work out what is wrong, I am not that familiar with ruby. Can you help at all? I have tried it on 3 systems, windows 7 and windows 10 with the same result.

ERROR:  Error installing stack_master:
        ERROR: Failed to build gem native extension.

```
current directory: N:/Ruby23/lib/ruby/gems/2.3.0/gems/gpgme-2.0.12/ext/gpgme
```

N:/Ruby23/bin/ruby.exe -r ./siteconf20161013-1180-gq47yd.rb extconf.rb

---

IMPORTANT!  gpgme gem uses locally built versions of required C libraries,
namely libgpg-error, libassuan, and gpgme.

If this is a concern for you and you want to use the system library
instead, abort this installation process and reinstall gpgme gem as
follows:

```
gem install gpgme -- --use-system-libraries
```

---

Extracting libgpg-error-1.21.tar.bz2 into tmp/i686-w64-mingw32/ports/libgpg-error/1.21... OK
Running 'configure' for libgpg-error 1.21... OK
Running 'compile' for libgpg-error 1.21... ERROR, review 'N:/Ruby23/lib/ruby/gems/2.3.0/gems/gpgme-2.0.12/ext/gpgme/tmp/
# i686-w64-mingw32/ports/libgpg-error/1.21/compile.log' to see what happened. Last lines are:

libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I.. -DLOCALEDIR=\"N:/Ruby23/lib/ruby/gems/2.3.0/gems/gpgme-2.0.12/ports/i686
-w64-mingw32/libgpg-error/1.21/share/locale\" -fPIC -Wall -Wpointer-arith -Wno-psabi -MT libgpg_error_la-estream.lo -MD
-MP -MF .deps/libgpg_error_la-estream.Tpo -c estream.c -o libgpg_error_la-estream.o
estream.c:1:0: warning: -fPIC ignored for target (all code is position independent) [enabled by default]
estream.c: In function 'es_fill':
estream.c:1714:24: error: 'EWOULDBLOCK' undeclared (first use in this function)
estream.c:1714:24: note: each undeclared identifier is reported only once for each function it appears in
estream.c: In function 'es_flush':
estream.c:1784:28: error: 'EWOULDBLOCK' undeclared (first use in this function)
estream.c: In function 'es_read_nbf':
estream.c:2044:24: error: 'EWOULDBLOCK' undeclared (first use in this function)
estream.c: In function 'es_seek':
estream.c:2316:20: error: 'EWOULDBLOCK' undeclared (first use in this function)
estream.c: In function 'es_write_nbf':
estream.c:2374:24: error: 'EWOULDBLOCK' undeclared (first use in this function)
make[3]: **\* [libgpg_error_la-estream.lo] Error 1
make[3]: Leaving directory `/n/Ruby23/lib/ruby/gems/2.3.0/gems/gpgme-2.0.12/ext/gpgme/tmp/i686-w64-mingw32/ports/libgpg-
error/1.21/libgpg-error-1.21/src'
make[2]: *** [all] Error 2
make[2]: Leaving directory`/n/Ruby23/lib/ruby/gems/2.3.0/gems/gpgme-2.0.12/ext/gpgme/tmp/i686-w64-mingw32/ports/libgpg-
error/1.21/libgpg-error-1.21/src'
make[1]: **\* [all-recursive] Error 1
make[1]: Leaving directory `/n/Ruby23/lib/ruby/gems/2.3.0/gems/gpgme-2.0.12/ext/gpgme/tmp/i686-w64-mingw32/ports/libgpg-
error/1.21/libgpg-error-1.21'
# make: **\* [all] Error 2

**\* extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
        --with-opt-dir
        --without-opt-dir
        --with-opt-include
        --without-opt-include=${opt-dir}/include
        --with-opt-lib
        --without-opt-lib=${opt-dir}/lib
        --with-make-prog
        --without-make-prog
        --srcdir=.
        --curdir
        --ruby=N:/Ruby23/bin/$(RUBY_BASE_NAME)
        --clean
        --use-system-libraries
N:/Ruby23/lib/ruby/gems/2.3.0/gems/mini_portile2-2.1.0/lib/mini_portile2/mini_portile.rb:366:in `block in execute': Fail
ed to complete compile task (RuntimeError)
        from N:/Ruby23/lib/ruby/gems/2.3.0/gems/mini_portile2-2.1.0/lib/mini_portile2/mini_portile.rb:337:in`chdir'
        from N:/Ruby23/lib/ruby/gems/2.3.0/gems/mini_portile2-2.1.0/lib/mini_portile2/mini_portile.rb:337:in `execute'
        from N:/Ruby23/lib/ruby/gems/2.3.0/gems/mini_portile2-2.1.0/lib/mini_portile2/mini_portile.rb:111:in`compile'
        from N:/Ruby23/lib/ruby/gems/2.3.0/gems/mini_portile2-2.1.0/lib/mini_portile2/mini_portile.rb:150:in `cook'
        from extconf.rb:76:in`block in <main>'
        from extconf.rb:65:in `tap'
        from extconf.rb:65:in`<main>'

extconf failed, exit code 1

Gem files will remain installed in N:/Ruby23/lib/ruby/gems/2.3.0/gems/gpgme-2.0.12 for inspection.
Results logged to N:/Ruby23/lib/ruby/gems/2.3.0/extensions/x86-mingw32/2.3.0/gpgme-2.0.12/gem_make.out

### PR DESCRIPTION
**Pull request by patrobinson**
Fixes #134 

This makes it possible to install StackMaster natively on Windows. It requires that when we release a Gem though that we build two versions, one on Mac (also works on Linux) and one on Windows. By shipping both versions to RubyGems, when a user runs `gem install stack_master` it will prefer one that is specifically built for their platform over the universal gem. If however we just release the universal gem, users on Windows will get the same error they do today and will have to run `gem install stack_master --platform x86-mingw32`.

I'm thinking of having some internal build servers at Envato do this for us, so we don't have to spin up a Windows machine just to build a Gem, but at the moment I am just doing this on a WorkSpace host.

EDIT:
In 2.0 we plan to remove dotgpg support for _something_ else. When that's done we can have the same gem for all platforms.

### CODE REVIEW

**Comment by patrobinson** [comment]
Looks like we have some code that isn't native either, I'll put this in WIP for now https://github.com/envato/stack_master/blob/master/lib/stack_master/commands/terminal_helper.rb#L5

**Comment by patrobinson** [comment]
The tests pass on Windows now, although Travis CI doesn't support it so you'll just have to take my word for it.

File: lib/stack_master/parameter_resolvers/secret.rb [hunk 1]
```diff
@@ -1,18 +1,23 @@
...
+        array_resolver  
+      end
 
       def initialize(config, stack_definition)
         @config = config
         @stack_definition = stack_definition
       end
 
       def resolve(value)
+        raise PlatformNotSupported, "Windows does not support the GPG Secret Parameter Resolver" if OS.windows?
```
Comment (stevehodgkiss) [inline lib/stack_master/parameter_resolvers/secret.rb]: Maybe reword this to "The GPG Secret Parameter Resolver does not support Windows" ?

**Comment by rbayerl** [comment]
Any thoughts on rewriting the GPG secret resolver to shell out instead of use GPGme? Then the resolver will work on all platforms and you can keep a single global release. Just a thought - I have no horse in this race 😄

**Comment by patrobinson** [comment]
Shelling out is a path to the dark side. If there's no other option I'd be OK but I suspect there are better options I have not explored yet.

**Comment by rbayerl** [comment]
I have no objection. It just seems like a lot of work setting up build servers and maintaining two different code paths just to continue supporting a resolver.

Another option would be to remove the secret resolver from stack_master itself, but give directions on how to incorporate it locally in a project for anyone who is using it.

Or you could cut a 2.x branch that removes the resolver and supports windows and maintain two versions. This also introduces complexity, though.

These are just ideas. What you have looks great...it just seems it's going to introduce the overhead of pipelines to release a new version.

**Comment by patrobinson** [comment]
@rbayerl oh yes, I added this to the PR description after the fact:

> In 2.0 we plan to remove dotgpg support for something else. When that's done we can have the same gem for all platforms.
